### PR TITLE
Refactor rules for function definitions and parameters

### PR DESCRIPTION
**Pull request by lysnikolaou**
This is still a work in progress. `pytest` passes, but I still get a SEGFAULT in `simpy_cpython`, which I can't understand and which I am looking into.

If anyone can have a look and maybe find the source of the SEGFAULT, it would be great! What I know so far is that the file `cpython/Lib/test/test__xxsubinterpreters.py` is the one that cannot be parsed.

Edit: The segfault is out of the way. This still remains a WIP though, as I still have to update the docs for it.

### CODE REVIEW

File: pegen/pegen.c [hunk 1]
```diff
@@ -757,3 +757,192 @@ map_targets_to_del_names(Parser *p, asdl_seq *seq)
...
+    }
+
+    return new_seq;
+}
+
+asdl_seq *
+_get_names(Parser *p, asdl_seq *names_with_defaults)
+{
+    int len = asdl_seq_LEN(names_with_defaults);
+    asdl_seq *seq = _Py_asdl_seq_new(len, p->arena);
```
Comment (pablogsal) [inline pegen/pegen.c]: Please, check the return value before populating the element

File: pegen/pegen.c [hunk 2]
```diff
@@ -757,3 +757,192 @@ map_targets_to_del_names(Parser *p, asdl_seq *seq)
...
+        asdl_seq_SET(seq, i, pair->arg);
+    }
+    return seq;
+}
+
+asdl_seq *
+_get_defaults(Parser *p, asdl_seq *names_with_defaults)
+{
+    int len = asdl_seq_LEN(names_with_defaults);
+    asdl_seq *seq = _Py_asdl_seq_new(len, p->arena);
```
Comment (pablogsal) [inline pegen/pegen.c]: Same here

File: pegen/pegen.c [hunk 3]
```diff
@@ -757,3 +757,192 @@ map_targets_to_del_names(Parser *p, asdl_seq *seq)
...
+
+arguments_ty
+make_arguments(Parser *p, asdl_seq *slash_without_default, SlashWithDefault *slash_with_default,
+               asdl_seq *plain_names, asdl_seq *names_with_default, StarEtc *star_etc)
+{
+    asdl_seq *posonlyargs;
+    if (slash_without_default != NULL) {
+        posonlyargs = slash_without_default;
+    } else if (slash_with_default != NULL) {
+        posonlyargs = _join_seqs(p,
```
Comment (pablogsal) [inline pegen/pegen.c]: `_join_seqs` and `_get_names`can fail, check the return value before passing the value arround

File: pegen/pegen.c [hunk 4]
```diff
@@ -757,3 +757,192 @@ map_targets_to_del_names(Parser *p, asdl_seq *seq)
...
+    }
+
+    arg_ty vararg = NULL;
+    if (star_etc != NULL && star_etc->vararg != NULL) {
+        vararg = star_etc->vararg;
+    }
+
+    asdl_seq *kwonlyargs;
+    if (star_etc != NULL && star_etc->kwonlyargs != NULL) {
+        kwonlyargs = _get_names(p, star_etc->kwonlyargs);
```
Comment (pablogsal) [inline pegen/pegen.c]: `_get_names` can fail.

File: pegen/pegen.c [hunk 5]
```diff
@@ -757,3 +757,192 @@ map_targets_to_del_names(Parser *p, asdl_seq *seq)
...
+        posargs = plain_names;
+    } else {
+        posargs = _Py_asdl_seq_new(0, p->arena);
+        if (!posargs)
+            return NULL;
+    }
+
+    asdl_seq *posdefaults;
+    if (slash_with_default != NULL && names_with_default != NULL) {
+        posdefaults = _join_seqs(p,
```
Comment (pablogsal) [inline pegen/pegen.c]: All of these calls can fail (`_join_seqs`, `_get_defaults` ).

**Comment by pablogsal** [comment]
The segfault is because of the result `AST`is malformed and there is an expression node with an invalid value for `_expr_kind`  (52 while the enum only has 26).

**Comment by lysnikolaou** [comment]
@pablogsal Thanks for the review. I will push a new commit ASAP. In the meantime, the segfault is due to the statement `raise ...`, which is not being parse correctly.

**Comment by lysnikolaou** [comment]
Would it be better to open a new PR for the fix or just push to this one? It has to do with the `raise_stmt`, but it is only a single change in the `atom` rule and the addition of a test file with `raise ...`.

**Comment by pablogsal** [comment]
> Would it be better to open a new PR for the fix or just push to this one?

I would say another PR, as this one is already big enough ;)

**Comment by gvanrossum** [comment]
If Pablo thinks this is ready, don't wait for me -- I have a few other things to take care of, and I find it hard to make time for substantial code reviews. Sorry!

**Comment by gvanrossum** [comment]
Yes, I was trying to follow PEP 7 (though maybe not hard enough :-).

**Comment by lysnikolaou** [comment]
@pablogsal Thanks a lot for the review!! This one definitely wasn't an easy one for me! :tada:

**Pull request merged**